### PR TITLE
Enforce unique group memberships for inserts/updates

### PR DESCRIPTION
**Pull request by nwalters512**
This is the first step towards fixing #8394.

I found places we were inserting into `group_users` with the following regexp:

```
INSERT INTO.*\n?.*group_users
```

I then updated all queries (or the surrounding JS) to perform checks to ensure that there's not already another row placing the user in another group.

The only `INSERT` I didn't have to touch was `assessment_groups_copy`, which deletes all existing groups before creating new ones.

### CODE REVIEW

File: apps/prairielearn/src/lib/groups.sql [hunk 1]
```diff
@@ -169,7 +169,8 @@ FROM
   JOIN group_users AS gu ON gu.group_id = g.id
 WHERE
   gc.assessment_id = $assessment_id
-  AND gu.user_id = $user_id;
+  AND gu.user_id = $user_id
+  AND g.deleted_at IS NULL;
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/groups.sql:173]: Groups are soft-deleted; I believe we want to disregard membership in deleted groups here.

File: apps/prairielearn/src/sprocs/assessment_groups_copy.sql [hunk 2]
```diff
@@ -32,14 +32,22 @@ BEGIN
...
-
+    WITH deleted_groups AS (
+      UPDATE groups g
+      SET deleted_at = NOW()
+      WHERE g.group_config_id = temp_old_group_config_id
+    )
     INSERT INTO group_logs 
-        (authn_user_id, group_id, action)
-    VALUES 
-        (assessment_groups_copy.authn_user_id, temp_old_group_config_id, 'delete');
```
Comment (nwalters512) [inline apps/prairielearn/src/sprocs/assessment_groups_copy.sql:42]: This feels just completely broken to me; we're inserting a `group_configs.id` value as `group_id`?

Given that this has already produced completely invalid data, I have half a mind to drop the `group_logs` table entirely (or rename it to `group_logs_old`) and start over from scratch. Not in this PR, of course.

Given _that_, maybe I don't bother making this change at all? 🤷

File: apps/prairielearn/src/tests/groupStudent.test.js [hunk 3]
```diff
@@ -339,6 +351,83 @@ describe('Group based homework assess control on student side', function () {
...
+      locals.$ = cheerio.load(page);
+    });
+    it('should contain the 4-character join code', function () {
+      const elemList = locals.$('#join-code');
+      locals.joinCode = elemList.text();
+      assert.lengthOf(locals.joinCode, locals.$('#group-name').text().length + 1 + 4);
+    });
+  });
+
+  describe('13.75. The first user cannot join the second group', () => {
```
Comment (nwalters512) [inline apps/prairielearn/src/tests/groupStudent.test.js:395]: "13.75" lmao. I resisted the urge to rewrite all these tests.

File: apps/prairielearn/src/tests/instructorAssessmentGroups.test.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+  step('cannot add a user to a group if they are already in another group', async () => {
+    const getResponse = await fetchCheerio(instructorAssessmentGroupsUrl, {});
+
+    // The add member form is dynamically rendered on the client, so we need to
+    // grab the CSRF token from somewhere else instead of getting it from the
+    // actual form.
+    const csrfToken = getCSRFToken(getResponse.$);
+
+    const response = await fetchCookie(fetchCheerio)(instructorAssessmentGroupsUrl, {
```
Comment (nwalters512) [inline apps/prairielearn/src/tests/instructorAssessmentGroups.test.ts:140]: I'm wondering if `fetchCheerio` should automatically use `fetchCookie` internally? When testing with flash messages, I repeatedly forgot to wrap the fetch with `fetchCookie` to make that work.

File: apps/prairielearn/src/lib/groups.ts [hunk 5]
```diff
@@ -196,8 +196,19 @@ export async function joinGroup(
     flash('error', 'The join code has an incorrect format');
     return;
   }
+
   const groupName = splitJoinCode[0];
   const joinCode = splitJoinCode[1].toUpperCase();
+
+  // This is a best-effort check to produce a nice error message. Even if this
+  // fails due to a race condition, the `group_users_insert` sproc below will
+  // validate that the user isn't already in a group.
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/groups.ts:205]: I used this pattern throughout this PR: optimistically try to detect multiple memberships, but still pessimistically enforce uniqueness via locking in sprocs.

File: apps/prairielearn/src/sprocs/assessment_groups_add_member.sql [hunk 6]
```diff
@@ -12,6 +12,7 @@ BEGIN
     -- ##################################################################
     -- verify the updating group belongs to the selected assessment
     -- then lock the group row
+    -- TODO: This sproc and associated UI doesn't actually respect maximum size constraints.
```
Comment (nwalters512) [inline apps/prairielearn/src/sprocs/assessment_groups_add_member.sql:15]: 🙃 I didn't necessarily want to address this here. This constraint checking isn't as important, as only instructors are currently able to make group membership exceed the max size.

File: apps/prairielearn/src/sprocs/assessment_groups_copy.sql (hunk 2 above)
Comment (nwalters512) [reply apps/prairielearn/src/sprocs/assessment_groups_copy.sql:42]: Tangent: as discussed on #6256, copying is just totally broken. We could consider entirely removing the copying functionality for now until we can rewrite it "correctly" and avoid this whole discussion?

Comment (nwalters512) [reply apps/prairielearn/src/sprocs/assessment_groups_copy.sql:42]: Actually, _because_ this code is so broken, it has never produced any logs 🙃

Comment (nwalters512) [reply apps/prairielearn/src/sprocs/assessment_groups_copy.sql:42]: #8910 will remove copying entirely, meaning we don't have to worry about this at all. Yay!

File: apps/prairielearn/src/models/user.ts [hunk 7]
```diff
@@ -8,6 +8,10 @@ export async function selectUserById(user_id: string): Promise<User> {
   return await queryRow(sql.select_user_by_id, { user_id }, UserSchema);
 }
 
+export async function selectUserByUid(uid: string): Promise<User> {
+  return await queryRow(sql.select_user_by_uid, { uid }, UserSchema);
+}
+
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/models/user.ts]: Addressed in 0d30a81.

File: apps/prairielearn/src/tests/helperClient.ts [hunk 8]
```diff
@@ -39,11 +39,22 @@ export async function fetchCheerio(
     // version of this code patched this so consumers can use it as "normal",
     // but the new function was created as not async. This behaviour is kept for
     // backwards compatibility.
-    text: { value: () => text },
+    text: { value: () => Promise.resolve(text) },
```
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/tests/helperClient.ts:38]: Addressed in 9efb1ba.

**Review by jonatanschroeder — APPROVED** [review]

File: apps/prairielearn/src/sprocs/assessment_groups_copy.sql (hunk 2 above)
Comment (nwalters512) [reply apps/prairielearn/src/sprocs/assessment_groups_copy.sql]: After the merge in 7037811, we no longer have to deal with this at all.

**Pull request merged**